### ISSUE TITLE
Using gRPC/protobuf 

### ISSUE THREAD

**Issue by anasalkouz** [opensearch-project/OpenSearch#1287]
Replace REST/JSON-based API with gRPC/protobuf to improve serialization/deserialization performance.
Ref: https://grpc.io/docs/what-is-grpc/introduction/

**Comment by AmiStrn** [comment]
Would this also help with streaming rather than doc by doc or bulk indexing?

**Comment by penghuo** [comment]
Would gRPC also benefit node-to-node communication?

**Comment by anasalkouz** [comment]
> Would gRPC also benefit node-to-node communication?

Sure, this should help in both node to node communication (Internal) and client to server communication (External).

**Comment by dblock** [comment]
@saratvemulapalli This becomes theoretically swappable via the extensions work, doesn't it?

**Comment by anasalkouz** [comment]
Based on my understanding of how the internal communication between the node works now, it's using a single transport layer and it should be easy to change how to works. @mch2 please keep me honest.

**Comment by saratvemulapalli** [comment]
@dblock Extensions are being built over transport (layer-4), GPRC/Protobuf operate atleast at layer 5. 
Extensions are adaptable to new communication mechanisms. 

@anasalkouz that is correct. [Netty4 transport](https://github.com/opensearch-project/OpenSearch/tree/main/modules/transport-netty4) is the module which implements transport in OpenSearch via NIO. This is swappable with new communication mechanisms. Netty4 also includes Layer 7 (HTTP) support.

**Comment by anasalkouz** [comment]
Shall we swap Netty4 with gPRC vs upgrading to [Netty 5](https://github.com/opensearch-project/OpenSearch/issues/2619)  ?

**Comment by dblock** [comment]
> Shall we swap Netty4 with gPRC vs upgrading to [Netty 5](https://github.com/opensearch-project/OpenSearch/issues/2619) ?

Yes! I would enable that as an option/experiment and benchmark it.

**Comment by saratvemulapalli** [comment]
Opened up a draft proposal for Protobuf in OpenSearch: https://github.com/opensearch-project/OpenSearch/issues/6844

**Comment by navneet1v** [comment]
@dblock is there a separate issue to track the gRPC protocol integration in OpenSearch? or it is already done?

**Comment by saratvemulapalli** [comment]
@navneet1v there isn't one. We could use this issue to track.
cc: @VachaShah

**Comment by navneet1v** [comment]
@saratvemulapalli if we are using this issue to track can we add more details in the description around what we are doing and all the tasks which are tracking for this.

**Comment by Bukhtawar** [comment]
Just so as to clarify we are only evaluating the data serialisation protocol(protobuf/avro/ion) etc and not really the transport layers like Netty/gRPC etc

**Comment by saratvemulapalli** [comment]
@navneet1v as @Bukhtawar said there is no body working on transport protocol yet (like GPRC). @VachaShah is exploring different serialization/de-serialization mechanisms we can adopt for OpenSearch. 

Do you see a use case for gRPC? What are you looking for ?

**Comment by navneet1v** [comment]
What I am looking for is a light weight protocol than HTTP. While doing a search we spend a significant amount to time at this HTTP layer so, my thinking was how we can replace this so that we can get better query latencies.

**Comment by Bukhtawar** [comment]
@navneet1v to clarify, we use TCP for node to node communication, did you mean HTTP at the REST layer?

**Comment by anirudha** [comment]
Ideally i would like to see support for Apache Arrow 0-copy via any http/2 transport

**Comment by anirudha** [comment]
@anasalkouz this way you remove the need for serDeSer but; this might need more extensive changes.
https://arrow.apache.org/docs/format/Flight.html

**Comment by Bukhtawar** [comment]
@anirudha Netty(default HTTP/transport) already supports zero-copy. Lets not complicate this issue further
